### ISSUE TITLE
IL2CPP Problems

### ISSUE THREAD

**Issue by Futurlabmatt** [Real-Serious-Games/C-Sharp-Promise#47]
Hi all, I've just started using Promises in Unity3d (2017.1) because they look.... promising. I've got a very small test scene running on PS4 compiling with il2cpp, stable .net 3.5, api compat .net 2.0 and no il2cpp optimisation set, the latest promise DLLs from here and I get errors at runtime with this code:

`public class PromiseTest : MonoBehaviour
{
    private void Awake()
    {
        DoThing().Then(b => UnityEngine.Debug.Log("Done: "+b)).Done();
    }

    private Promise<bool> DoThing()
    {
        Promise<bool> p = new Promise<bool>();
        p.Resolve(true);
        return p;
    }

}`

Error at runtime:
`MissingMethodException: No constructor found for System.Runtime.CompilerServices.StrongBox`1[[RSG.Promise`1+<>c__DisplayClass6`1[[System.Boolean, mscorlib, Version=2.0.0.0, Culture=, PublicKeyToken=b77a5c561934e089],[System.Boolean, mscorlib, Version=2.0.0.0, Culture=, PublicKeyToken=b77a5c561934e089]], RSG.Promise, Version=1.1.2.0, Culture=, PublicKeyToken=null]]::.ctor(RSG.Promise`1+<>c__DisplayClass6`1[System.Boolean,System.Boolean])`

I've tried adding 
`<linker>
     <assembly fullname="System">	
		<namespace fullname="System.Runtime.CompilerServices" preserve="all"/>
	 </assembly>	
	 <assembly fullname="RSG.Promise">	
		<namespace fullname="RSG.Promise" preserve="all"/>
	 </assembly>
	 <assembly fullname="RSG.Toolkit">	
		<namespace fullname="RSG.Utils" preserve="all"/>
	 </assembly>	 
</linker>`

to a link.xml file, but still get the same problem. Does anyone know how this might be solved? I've love to use promises for my project.

**Comment by RoryDungan** [comment]
Hey, thanks for finding that issue. No one at Real Serious Games has a PS4 dev kit but I attempted to reproduce it using IL2CPP for WebGL. Unfortunately that seemed to work fine with your code so it may be a PS4-specific issue, not a general IL2CPP issue. 

The only time I've seen that error before is trying to use certain types of reflection on IL2CPP platforms. C-Sharp-Promise does not use reflection but RSG.Toolkit does, so do you think you could try it again with just the `RSG.Promise.dll` DLL in your project and none of the others? RSG.Toolkit is actually only a dependency of the tests and examples, not the core library, so you don't need to include it in your project.

If that works then I will probably just remove RSG.Toolkit from the project because it is only needed for a very small function anyway.

**Comment by Futurlabmatt** [comment]
Ah, okay, cool. I got this working using the suggestion above.

1. I removed the RSG.Toolkit.dll from the unity project
2. This caused some issues where some extensions references weren't available (seems there is a decependency)
3. In the promises VS solution, I removed the references of the toolkit from the examples folder projects to test I was properly removing the need for the toolkit reference
4. I added the entire LinqExts code from the RSG.Toolkit.dll->RSG.Utils.LinqExts in to the RSG.Promises code
5. I made a build of the new RSG.Promise.dll (now with linqExts) and added back to Unity
6. Build works, I'm testing more now.
7. There seems to be an issue with existing, previously working code I had (to do with a promise<list<type>>, but I can try and work this out later.

[promises.patch.txt](https://github.com/Real-Serious-Games/C-Sharp-Promise/files/1390602/promises.patch.txt)

Awesome!
